src/verifast1.ml

```diff
@@ -2785,6 +2841,96 @@ module VerifyProgram1(VerifyProgramArgs: VERIFY_PROGRAM_ARGS) = struct
       | Int (Unsigned, n1), Int (Signed, n2) -> if n2 <= n1 then t1 else t2
       | Int (s, n1), Int (_, n2) -> Int (s, max n1 n2)
 
+  open Java_Inference
+
+  let rec super_types (ObjType (cn, targs)) =
+    let supers =
+      if cn = "java.lang.Object" then
+        []
+      else
+        direct_supers (ObjType (cn, targs))
+    in
+    flatmap (fun s -> s::super_types s) supers
+
+  (* calculates the shared supertype that is more specific than any other shared supertype *)
+  let least_upper_bound l tps =
+    match tps with
+      [t] -> [t]
+    | [] -> failwith "calculating least upper bound of no type"
+    | tps -> 
+      let st_map = List.map (fun tp -> (tp, tp::super_types tp)) tps in
+      let filtered_candidate_set = (*intersection of all est sets *)
+        let intersection setA setB = List.filter (fun memA -> List.mem memA setB) setA in
+        let rec iter sets iset = match sets with
+          [] -> []
+        | [last] -> intersection iset last
+        | hd::tl -> iter tl (intersection iset hd)
+        in
+        iter (List.map snd st_map) (snd (List.hd st_map))
+      in
+      (* MEC = { V | V in EC, and for all W ≠ V in EC, it is NOT the case that W <: V } *)
+      let mec = filtered_candidate_set |> List.filter begin fun v -> (* V in EC *)
+        let children = filtered_candidate_set |> List.filter begin fun w ->  (* for every V, select all it's children in EC that are not V *)
+          if (w = v) then   
+            false
+          else
+            (is_subtype_of_ w v)
+        end
+        in (* if V has no children, it is the most specific candidate*)
+        children = []
+      end
+      in
+      mec 
+
+  let infer_type l unknownTypes knownTypes =
+    let rec inference_set set actual expected = 
+      match (actual, expected) with 
+        (hd::tl, hd0::tl0) -> begin match (hd,hd0) with
+        | (ObjType (t, targs), ObjType (t0, targs0)) when t != "null" && t0 != "null" ->
+          if proper_type hd && proper_type hd0 then
+            inference_set set tl tl0
+          else if List.length targs != List.length targs0 then
+            inference_error l ("Type mismatch during inference, amount of type arguments does not match. Unknown type: " ^ string_of_type hd ^ ". Matching type: " ^ string_of_type hd0 ^ ".")
+          else if t != t0 then
+            inference_error l ("Type mismatch during inference. Unknown type: " ^ string_of_type hd ^ ". Matching type: " ^ string_of_type hd0 ^ ".")

```

Comment:
So you do not support parameter type `List<T>` and argument type `ArrayList<Person>`?

Reply:
For some reason I thought this would require dependencies, but it does not. This case has been added  to the code and the test cases for method inference.